app/features/signup/verify-email.ts

```diff
@@ -0,0 +1,65 @@
+import { useEffect, useRef, useState } from 'react';
+import { useUserStore } from '~/features/user/user-provider';
+import { useToast } from '~/hooks/use-toast';
+
+export const useVerifyEmailOnLoad = ({
+  code,
+  onFailed,
+}: {
+  code: string | undefined;
+  onFailed: () => void;
+}) => {
+  const verifyEmail = useUserStore((state) => state.verifyEmail);
+  const { toast } = useToast();
+  const failedRef = useRef(onFailed);
+
+  useEffect(() => {
+    if (code) {
+      const handleVerificationCode = async () => {
+        try {
+          await verifyEmail(code);
+        } catch {
+          toast({
+            variant: 'destructive',
+            title: 'Failed to verify',
+            description:
+              'Please try entering the code manually or request another verification email',
+          });
+          failedRef.current();
```

Comment:
why use a ref here? Does this make it so that you don't have to put onFailed in dependency array of the useEffect, therefore if the `onFailed` function changes the effect does not run again?

Reply:
exactly